packages/yarnpkg-cli/sources/main.ts

```diff
@@ -77,7 +77,14 @@ export async function main({binaryVersion, pluginConfiguration}: {binaryVersion:
     const ignorePath = configuration.get(`ignorePath`);
     const ignoreCwd = configuration.get(`ignoreCwd`);
 
-    if (yarnPath !== null && !ignorePath) {
+    // Avoid unnecessary spawn when run directly
+    if (!ignorePath && !ignoreCwd && yarnPath === npath.toPortablePath(npath.resolve(process.argv[1]))) {
+      process.env.YARN_IGNORE_PATH = `1`;
+      process.env.YARN_IGNORE_CWD = `1`;
+
+      await exec(cli);
+      return;
+    } else if (yarnPath !== null && !ignorePath) {
```

Comment:
Why not just this (in particular I'm not sure about the `YARN_IGNORE_CWD` - what if the yarnPath is right, but there's a `--cwd` option)?

```suggestion
    if (yarnPath !== null && !ignorePath && yarnPath !== npath.toPortablePath(npath.resolve(process.argv[1]))) {
```

Reply:
I agree with the `ignoreCwd` part but the `configuration` isn't fully populated when `ignorePath` isn't set

Reply:
Perhaps there's a bug in the original code, because the flow doesn't make sense: if there's a yarnPath and a cwd, we're spawning a new process with the yarnPath (that's ok), but we also set ignoreCwd, causing it to be ignored (that's weird) 🤔

cc @paul-soporan since part of this code comes from #1134, I might be missing something

Reply:
I honestly don't remember why I did this in #1134 (and I can't see how it could possibly make sense), I *think* it was because the `process.chdir()` code triggered too many times when we didn't ignore the cwd, but I have no idea what's exactly going on there. I think this behavior could be seen by adding a `--cwd` test and checking if it breaks without the `YARN_IGNORE_CWD` 🤔

Reply:
I'd argue that is a seperate issue from what this PR is trying to fix and thus should be fixed in a seperate PR

Reply:
Technically yes, but it feels weird to merge a new codepath that mimics one that we know is broken 😅 I'd tend to prefer to fix both in one pass, but I'll greenlight the PR and let you decide.